R/data_duplicated.R

```diff
@@ -0,0 +1,39 @@
+#' @title Extract all duplicates
+#'
+#' @description Extract all duplicates, for visual inspection.
+#' Note that it also contains the first occurrence of future
+#' duplicates, unlike [duplicated()] or [dplyr::distinct()]). Also
+#' contains an additional column reporting the number of missing
+#' values for that row, to help in the decision-making when
+#' selecting which duplicates to keep.
+#'
+#' @param data The data frame.
+#' @param id The ID variable for which to check for duplicates.
+#' @keywords duplicates
+#' @export
+#' @return A dataframe, containing all duplicates.
+#' @examples
+#' df1 <- data.frame(
+#'    id = c(1, 2, 3, 1, 3),
+#'    item1 = c(NA, 1, 1, 2, 3),
+#'    item2 = c(NA, 1, 1, 2, 3),
+#'    item3 = c(NA, 1, 1, 2, 3)
+#' )
+#'
+#' data_duplicated(df1, id = "id")
+#'
+#' # Filter to exclude duplicates
+#' df2 <- df1[-c(1, 5),]
+#' df2
+
+data_duplicated <- function(data, id) {
+  Row <- seq(nrow(data))
+  data <- cbind(Row, data)
+  dups.index <- data[[id]] %in% data[[id]][duplicated(data[id])]
```

Comment:
This works only if `id` is one variable, right? I have cases where I want to check duplicated observations by country and year, but I guess I can't pass `data_duplicated(data, id = c("country", "year"))`

Reply:
You're correct. I had only thought about the scenario of using a single ID variable (thus why I first used `id` instead of `select`). Not sure how I will need to change the code to accommodate that, but `dplyr::distinct` does allow several arguments:

``` r
devtools::load_all("D:/github/forks/datawizard")
#> ℹ Loading datawizard

df1 <- data.frame(
  country = c("US", "CAN", "US", "US", "US"),
  year = c(2022, 2022, 2000, 2022, 2022),
  item1 = c(NA, 1, 1, 2, 3),
  item2 = c(NA, 1, 1, 2, 3),
  item3 = c(NA, 1, 1, 2, 3)
)
df1
#>   country year item1 item2 item3
#> 1      US 2022    NA    NA    NA
#> 2     CAN 2022     1     1     1
#> 3      US 2000     1     1     1
#> 4      US 2022     2     2     2
#> 5      US 2022     3     3     3

dplyr::distinct(df1, country, .keep_all = TRUE)
#>   country year item1 item2 item3
#> 1      US 2022    NA    NA    NA
#> 2     CAN 2022     1     1     1

dplyr::distinct(df1, country, year, .keep_all = TRUE)
#>   country year item1 item2 item3
#> 1      US 2022    NA    NA    NA
#> 2     CAN 2022     1     1     1
#> 3      US 2000     1     1     1

# Doesn't work with data_unique so far
data_unique(df1, id = c("country", "year"), method = "first")
#> Error in .subset2(x, i, exact = exact): subscript out of bounds
```

<sup>Created on 2022-11-06 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
The easiest way is probably to concatenate the id columns together to have a single "temporary_id" column on which you check duplicates, and you remove the "temporary_id" column at the end

Reply:
Thanks. That makes sense. Sometimes with creating temporary columns I'm afraid of overwriting a user column also called something similar like "temporary_id". Is there any sort of convention for this? Of course the weirder the added column name, the less likely it is to overwrite, but still not impossible.

Reply:
Right, usually I use "temporary_id" but it's true that it can be an existing name so I should change this habit...

You can create a random name in the function with `new_name <- sample(letters)`, and then use the value of `new_name` as the temporary id.

Here's a way to concatenate columns:
https://github.com/easystats/datawizard/blob/main/R/data_to_wide.R#L148

Reply:
Also, not sure if it would be worth pulling the `count_na` function out of `data_duplicated`, and if so, whether it should go to `datawizard` or `insight`.

Reply:
Instead of generating a random name every time the function is run, I think we should agree on a specific temporary column name and use that consistently within `easystats`.

Reply:
And do you want me to use the argument `select` instead of `id` (like in your original comment in the discussion), or like in the link you provided, `id_cols`? (thanks btw, it's useful)

Same for `method` vs `keep`.

Reply:
Yes I think you should use `select`. `id_cols` was just to match the arguments of `tidyr::pivot_` functions. I also prefer `keep`

Reply:
I now realize that your idea to use a randomly generated name might be better because otherwise one function using the same name will remove that column from the other function using it (as is happening here). However, using a random name creates problems with `data_filter` later on, so not quite sure of the best way forward yet.